klog/klog.c

```diff
@@ -0,0 +1,387 @@
+}
+
+
+static int klog_readerAdd(unsigned pid, unsigned nonblocking)
+{
+	klog_reader_t *r;
+
+	if (klog_readerFind(pid) != NULL)
+		return -EINVAL;
+
+	r = vm_kmalloc(sizeof(klog_reader_t));
+	if (r == NULL)
+		return -ENOMEM;
+
+	hal_memset(r, 0, sizeof(klog_reader_t));
+
+	r->nonblocking = !!nonblocking;
+	r->pid = pid;
+
+	proc_lockSet(&klog_common.lock);
+	r->ridx = klog_common.head;
+	LIST_ADD(&klog_common.readers, r);
+	proc_lockClear(&klog_common.lock);
+
+	return 0;
+}
+
+
+static ssize_t _klog_readln(klog_reader_t *r, char *buf, size_t sz)
+{
+	ssize_t n = 0;
+
+	while (r->ridx < klog_common.tail && n < sz) {
+		buf[n++] = _fifo_get(r->ridx++);
+		if (buf[n - 1] == '\n' || buf[n - 1] == '\0')
+			break;
+	}
+
+	/* Always end with newline */
+	if (n > 0 && buf[n - 1] != '\n') {
+		if (buf[n - 1] == '\0') {
+			buf[n - 1] = '\n';
+		}
+		else if (n < sz) {
+			buf[n++] = '\n';
+		}
+		else {
+			buf[n - 1] = '\n';
+			r->ridx--;
+		}
+	}
+
+	return n;
+}
+
+
+static ssize_t klog_read(klog_reader_t *r, char *buf, size_t sz)
+{
+	ssize_t ret;
+
+	proc_lockSet(&klog_common.lock);
+	/* We need to catch up the ring buffer's head */
+	if (r->ridx < klog_common.head) {
+		ret = -EPIPE;
+		r->ridx = klog_common.head;
+	}
+	else {
+		ret = _klog_readln(r, buf, sz);
+	}
+	proc_lockClear(&klog_common.lock);
+
+	return ret;
+}
+
+
+static void _klog_msgRespond(klog_reader_t *r, ssize_t err)
+{
+	klog_readMsg_t *rmsg;
+	msg_t msg;
+
+	rmsg = r->msgs;
+	LIST_REMOVE(&r->msgs, rmsg);
+
+	msg.i.data = NULL;
+	msg.i.size = 0;
+
+	msg.type = mtRead;
+	msg.pid = r->pid;
+	msg.o.data = rmsg->odata;
+	msg.o.size = rmsg->osize;
+	msg.o.io.err = err;
+
+	proc_respond(klog_common.port, &msg, rmsg->rid);
+
+	vm_kfree(rmsg);
+}
+
+
+static void _klog_updateReaders(void)
+{
+	klog_reader_t *r = klog_common.readers;
+	ssize_t ret;
+
+	if (r != NULL) {
+		do {
+			if (r->msgs != NULL) {
+				ret = _klog_readln(r, r->msgs->odata, r->msgs->osize);
+				_klog_msgRespond(r, ret);
+			}
+			r = r->next;
+		} while (r != klog_common.readers);
+	}
+}
+
+
+static void klog_close(unsigned pid)
+{
+	klog_reader_t *r;
+
+	proc_lockSet(&klog_common.lock);
+	r = klog_readerFind(pid);
+	if (r != NULL) {
+		while (r->msgs != NULL)
+			_klog_msgRespond(r, -EIO);
+		LIST_REMOVE(&klog_common.readers, r);
+		vm_kfree(r);
+	}
+	proc_lockClear(&klog_common.lock);
+}
+
+
+static int klog_devctl(msg_t *msg)
+{
+	ioctl_in_t *in;
+	ioctl_out_t *out;
+
+	in = (ioctl_in_t *)msg->i.raw;
+	out = (ioctl_out_t *)msg->o.raw;
+	/*
+	 * We need to handle isatty(), which
+	 * only checks if a device responds to TCGETS
+	 */
+	if (in->request == TCGETS) {
+		out->err = EOK;
+	}
+	else {
+		out->err = -EINVAL;
+	}
+
+	return 0;
+}
+
+
+int klog_write(const char *data, size_t len)
+{
+#if KLOG_ENABLE
+	int i = 0, overwrite = 0;
+	char c;
+
+	proc_lockSet(&klog_common.lock);
+
+	if (klog_common.tail + len >= klog_common.head + KLOG_BUFSZ)
+		overwrite = 1;
+
+	while (i < len)
+		_fifo_push(data[i++]);
```

Comment:
Nitpick - for fits better
```suggestion
	for (i = 0; i < len; ++i)
		_fifo_push(data[i]);
```